src/Collection/Iterator/UniqueIterator.php

```diff
@@ -0,0 +1,45 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         3.0.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Collection\Iterator;
+
+class UniqueIterator extends FilterIterator
+{
+    protected array $_exists = [];
+
+    /**
+     * @param \Traversable|array $items
+     * @param string|null $key
+     */
+    public function __construct($items, ?string $key)
+    {
+        parent::__construct($items, function($item) use ($key) {
+
+            if (is_null($key)) {
+                $key = $value = $item;
+            } else {
+                $value = $item[$key] ?? null;
+            }
+
+            if(isset($this->_exists[$value])) {
```

Comment:
It seems like `$value` could be a value that isn't suitable for an array key. Examples that come to mind are datetime values or associated records. Should we also have a check here to ensure that the value can be used as array key, or perhaps have a different solution that will work for all value types?

Reply:
Yes, you are right. I made appropriate changes. What do you think about implementation?

P.S. I will fix all CS issues when we finish the final implementation.